### PR TITLE
lib: Add optionalDrvAttr to conditionally set drv attributes.

### PR DESCRIPTION
**Pull request by shlevy**
This allows for adding new, conditionally set, derivation attributes to an existing derivation without changing any output paths in the case where the condition is not met.

## Things done

- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by shlevy** [comment]
@infinisil any concern?

File: lib/derivations.nix [hunk 1]
```diff
@@ -98,4 +98,30 @@ in
...
     } // passthru;
+
+  /* Conditionally set a derivation attribute.
+
+     Because `mkDerivation` sets `__ignoreNulls = true`, a derivation
+     attribute set to `null` will not impact the derivation output hash.
+     Thus, this function passes through its `value` argument if the `cond`
+     is `true`, but returns `null` if not.
+
+     Type: optionalDrvAttr :: Bool -> a -> a?
```
Comment (infinisil) [inline lib/derivations.nix:109]:
```suggestion
     Type: optionalDrvAttr :: Bool -> a -> a | Null
```

**Review by infinisil — APPROVED** [review]
This name is much better than the original `optionalString'` or the other proposals in [the thread](https://github.com/NixOS/nixpkgs/pull/284874#discussion_r1471866439). Only something minor, otherwise this LGTM!

File: lib/derivations.nix (hunk 1 above)
Comment (shlevy) [reply lib/derivations.nix]: Fixed

**Pull request merged**